### ISSUE TITLE
`http::uri::Uri` is excessively large and makes unnecessary allocations

### ISSUE THREAD

**Issue by K4rakara** [hyperium/http#687]
`http::Uri` contains the following:
```rust
pub struct Uri {
    scheme: Scheme,
    authority: Authority,
    path_and_query: PathAndQuery,
}
```
Makes sense, simple and terse.
If we take a look at the `http::uri::scheme::Scheme` struct, we get the following:
```rust
#[derive(Clone)]
pub struct Scheme {
    pub(super) inner: Scheme2,
}

#[derive(Clone, Debug)]
pub(super) enum Scheme2<T = Box<ByteStr>> {
    None,
    Standard(Protocol),
    Other(T),
}

#[derive(Copy, Clone, Debug)]
pub(super) enum Protocol {
    Http,
    Https,
}
```

Here is where we start running into problems. At a first glance, this might make sense. By having it be an enum with three variants, you can optimize the most common usecases while still supporting others, and by having the `Other` variant store a boxed value by default, you can make the enum just five bytes wide. `Bytes` is 32 bytes wide (on x86_64), so the boxing _almost_ makes sense here. I'll explain in a moment why it doesn't.

Next, lets look at the `http::uri::authority::Authority` struct:
```rust
#[derive(Clone)]
pub struct Authority {
    pub(super) data: ByteStr,
}
```

This struct wraps another `ByteStr`, which wraps a `Bytes` value. Not the end of the world.

Finally, lets look at the `http::uri::path::PathAndQuery` struct:
```rust
#[derive(Clone)]
pub struct PathAndQuery {
    pub(super) data: ByteStr,
    pub(super) query: u16,
}
```

This struct uses yet another `ByteStr`, which is really unnecessary. We already have at least one `ByteStr` handle, which in and of itself is 32 bytes. There's no real reason to have multiple of them, when you can instead store an index into the `ByteStr`, like this struct actually does for the query portion.

I think it would be advisable if the layout of `Uri` was changed to resemble how `PathAndQuery` stores the index of the query, but with all the URI components. I'd estimate that you could get `Uri` to 44 to 56 bytes if you were to do this, and it wouldn't need to make any allocations.

# TL;DR

`http::uri::Uri` is **88 bytes wide**, before even accounting for any heap allocations it points to. This is because it holds several `ByteStr`s, which wrap `Bytes`, which are 32 bytes apiece (on x86_64).

# What Should be Done Instead?

Store a single `ByteStr` value in `Uri`, as well as indices for the various components of the URI, possibly as `u16`s, like the `url` crate does.  The various auxiliary structs can stay, and be created on request based on the stored indices, rather than storing them directly, which is costly.

**Comment by K4rakara** [comment]
now realizing that `url::Url` is 88 bytes in size... apologies. I wrote this up at 1am in my timezone, and it shows.

**Comment by seanmonstar** [comment]
I mean, the issue isn't crazy. I've desired before to reduce the size, too. I'm not entirely sure how best to do, though.

Just to answer one point, one of the reasons that the `Uri` pieces have their own `Bytes`, instead of a single one with indices, is to support HTTP/2 and 3. They receive the request-target as 3 separate headers, so to combine them into one would require creating a new string and copying into it. :shrug:
